libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/InitialGpsEventFactoryTest.java

```diff
@@ -0,0 +1,104 @@
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyZeroInteractions;
+
+public class InitialGpsEventFactoryTest {
+
+  @Test
+  public void navigationStarted_elapsedTimeStarts() {
+    ElapsedTime time = mock(ElapsedTime.class);
+    InitialGpsEventHandler handler = mock(InitialGpsEventHandler.class);
+    InitialGpsEventFactory factory = new InitialGpsEventFactory(time, handler);
+
+    factory.navigationStarted("some_session");
+
+    verify(time).start();
+  }
+
+  @Test
+  public void gpsReceived_elapsedTimeEnds() {
+    ElapsedTime time = mock(ElapsedTime.class);
+    InitialGpsEventHandler handler = mock(InitialGpsEventHandler.class);
+    InitialGpsEventFactory factory = new InitialGpsEventFactory(time, handler);
+
+    factory.gpsReceived();
+
+    verify(time).end();
+  }
+
+  @Test
+  public void validData_sendsCorrectEvent() {
+    ElapsedTime time = mock(ElapsedTime.class);
+    InitialGpsEventHandler handler = mock(InitialGpsEventHandler.class);
+    String sessionId = "some_session";
+    InitialGpsEventFactory factory = new InitialGpsEventFactory(time, handler);
+
+    factory.navigationStarted(sessionId);
+    waitingForGps();
+    factory.gpsReceived();
+
+    verify(handler).send(anyDouble(), eq(sessionId));
+  }
+
+  @Test
+  public void validData_doesNotSendEventTwice() {
+    ElapsedTime time = mock(ElapsedTime.class);
+    InitialGpsEventHandler handler = mock(InitialGpsEventHandler.class);
+    String sessionId = "some_session";
+    InitialGpsEventFactory factory = new InitialGpsEventFactory(time, handler);
+
+    factory.navigationStarted(sessionId);
+    waitingForGps();
+    factory.gpsReceived();
+    factory.gpsReceived();
+
+    verify(handler, times(1)).send(anyDouble(), eq(sessionId));
+  }
+
+  @Test
+  public void invalidStart_doesNotSend() {
+    ElapsedTime time = mock(ElapsedTime.class);
+    InitialGpsEventHandler handler = mock(InitialGpsEventHandler.class);
+    InitialGpsEventFactory factory = new InitialGpsEventFactory(time, handler);
+
+    factory.gpsReceived();
+
+    verifyZeroInteractions(handler);
+  }
+
+  @Test
+  public void reset_allowsNewEventToBeSent() {
+    ElapsedTime time = mock(ElapsedTime.class);
+    InitialGpsEventHandler handler = mock(InitialGpsEventHandler.class);
+    String firstSessionId = "first_session";
+    String secondSessionId = "second_session";
+    InitialGpsEventFactory factory = new InitialGpsEventFactory(time, handler);
+
+    factory.navigationStarted(firstSessionId);
+    waitingForGps();
+    factory.gpsReceived();
+    factory.gpsReceived();
+    factory.reset();
+    factory.navigationStarted(secondSessionId);
+    waitingForGps();
+    factory.gpsReceived();
+
+    verify(handler, times(1)).send(anyDouble(), eq(firstSessionId));
+    verify(handler, times(1)).send(anyDouble(), eq(secondSessionId));
+  }
+
+  private void waitingForGps() {
+    try {
+      Thread.sleep(1500);
```

Comment:
We should reconsider using `sleep` in tests 👀 https://github.com/twitter/commons/blob/master/src/java/com/twitter/common/styleguide.md#threadsleep

> **Thread.sleep()**
>
> Sleeping is rarely warranted, especially in test code. Sleeping is expressing an expectation that something else is happening while the executing thread is suspended. This quickly leads to brittleness; for example if the background thread was not scheduled while you were sleeping.
>
> Sleeping in tests is also bad because it sets a firm lower bound on how fast tests can execute. No matter how fast the machine is, a test that sleeps for one second can never execute in less than one second. Over time, this leads to very long test execution cycles.

Especially when writing _Behavior Verification_ tests in which we're not 👀 concrete state.

Reply:
@Guardiola31337 removed the sleep for an empty operation 👍